File: juno/strategies/rsi.py
from decimal import Decimal

from juno import Candle, CandleMeta, indicators
from juno.constraints import Int, Uniform

from .strategy import Oscillator, Strategy


# TODO: For example, well-known market technician Constance Brown, CMT, has promoted the idea that
# an oversold reading on the RSI in an uptrend is likely much higher than 30%, and an overbought
# reading on the RSI during a downtrend is much lower than the 70% level.
class Rsi(Oscillator):
    @staticmethod
    def meta() -> Strategy.Meta:
        return Strategy.Meta(
            constraints={
                "period": Int(1, 101),
                "up_threshold": Uniform(Decimal("50.0"), Decimal("100.0")),
                "down_threshold": Uniform(Decimal("0.0"), Decimal("50.0")),
            }
        )

    indicator: indicators.Rsi
    _up_threshold: Decimal
    _down_threshold: Decimal

    def __init__(
        self,
        period: int = 14,
        up_threshold: Decimal = Decimal("70.0"),
        down_threshold: Decimal = Decimal("30.0"),
    ) -> None:
        assert period > 0
        assert up_threshold >= down_threshold

        self.indicator = indicators.Rsi(period)
        self._up_threshold = up_threshold
        self._down_threshold = down_threshold

    @property
    def maturity(self) -> int:
        return self.indicator.maturity

    @property
    def mature(self) -> bool:
        return self.indicator.mature

    @property
    def overbought(self) -> bool:
        return self.indicator.mature and self.indicator.value >= self._up_threshold

    @property
    def oversold(self) -> bool:
        return self.indicator.mature and self.indicator.value < self._down_threshold

    def update(self, candle: Candle, _: CandleMeta) -> None:
        self.indicator.update(candle.close)

File: juno_rs/src/strategies/rsi.rs
use super::{Oscillator, Strategy, StrategyMeta};
use crate::{genetics::Chromosome, indicators, Candle};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Chromosome, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct RsiParams {
    pub period: u32,
    pub up_threshold: f64,
    pub down_threshold: f64,
}

fn period(rng: &mut StdRng) -> u32 {
    rng.gen_range(1..101)
}
fn up_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(50.0..100.0)
}
fn down_threshold(rng: &mut StdRng) -> f64 {
    rng.gen_range(0.0..50.0)
}

pub struct Rsi {
    indicator: indicators::Rsi,
    up_threshold: f64,
    down_threshold: f64,
}

impl Rsi {
    pub fn new(params: &RsiParams, _meta: &StrategyMeta) -> Self {
        Self {
            indicator: indicators::Rsi::new(params.period),
            up_threshold: params.up_threshold,
            down_threshold: params.down_threshold,
        }
    }
}

impl Strategy for Rsi {
    fn maturity(&self) -> u32 {
        self.indicator.maturity()
    }

    fn mature(&self) -> bool {
        self.indicator.mature()
    }

    fn update(&mut self, candle: &Candle) {
        self.indicator.update(candle.close);
    }
}

impl Oscillator for Rsi {
    fn overbought(&self) -> bool {
        self.indicator.mature() && self.indicator.value >= self.up_threshold
    }

    fn oversold(&self) -> bool {
        self.indicator.mature() && self.indicator.value < self.down_threshold
    }
}
